eskip/eskip.go

```diff
@@ -185,6 +179,85 @@ type RouteInfo struct {
 	ParseError error
 }
 
+// HasPredicate checks if the route has predicate with the given name.
+func (r *Route) HasPredicate(name string) bool {
+	for _, p := range r.Predicates {
+		if p.Name == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+// GetPredicates returns all predicates that match the given name.
+func (r *Route) GetPredicates(name string) []*Predicate {
+	pp := make([]*Predicate, 0)
+	for _, p := range r.Predicates {
+		if p.Name == name {
+			pp = append(pp, p)
+		}
+	}
+
+	return pp
+}
+
+// GetPredicate returns matching predicate or error when multiple predicates are given.
+func (r *Route) GetPredicate(name string) (*Predicate, error) {
+	pp := make([]*Predicate, 0)
+	for _, p := range r.Predicates {
+		if p.Name == name {
+			pp = append(pp, p)
+		}
+	}
+
+	if len(pp) == 0 {
+		return nil, nil
+	}
+
+	if len(pp) > 1 {
+		return nil, fmt.Errorf("multiple predicates of the same name: %d %s", len(pp), name)
+	}
+
+	return pp[0], nil
+}
+
+func (r *Route) PrependPredicate(pp ...*Predicate) error {
+	for _, p := range pp {
+		switch p.Name {
+		case "Path", "Weight":
+			if r.HasPredicate(p.Name) == true {
+				return fmt.Errorf("predicate of type %s can only be added once", p.Name)
+			}
+		}
+
+		r.Predicates = append([]*Predicate{p}, r.Predicates...)
+	}
+
+	return nil
+}
+
+func (r *Route) AppendPredicate(pp ...*Predicate) error {
+	for _, p := range pp {
+		switch p.Name {
+		case "Path", "Weight":
+			if r.HasPredicate(p.Name) == true {
+				return fmt.Errorf("predicate of type %s can only be added once", p.Name)
+			}
+		}
+
+		r.Predicates = append(r.Predicates, p)
+	}
+
+	return nil
+}
+
+func (r *Route) ReplacePredicates(pp ...*Predicate) error {
+	r.Predicates = make([]*Predicate, 0, len(pp))
```

Comment:
Shouldn't we first check whether there is an error, and only then replace the existing predicates. To my that would make more sense. With the proposed way of making PredicatesByName, this could become:

```
if len(PredicatesByName(pp, "Path")) > 1 ||
    len(PredicatesByName(pp, "PathSubtree")) > 1 ||
    len(PredicatesByName(pp, "Weight")) > 1 {
    return fmt.Errorf()
}

r.Predicates = pp
return nil
```